src/components/ha-target-picker.ts

```diff
@@ -250,11 +251,10 @@ export class HaTargetPicker extends SubscribeMixin(LitElement) {
 
   private async _showPicker(ev) {
     this._addMode = ev.currentTarget.type;
-    await this.updateComplete;
-    setTimeout(() => {
-      this._inputElement?.open();
-      this._inputElement?.focus();
-    }, 0);
+    await nextRender();
+    this._inputElement?.open();
+    await nextRender();
+    this._inputElement?.focus();
```

Comment:
I think this should be:
```ts
    await this._inputElement?.focus();
    this._inputElement?.open();
```

With some updates to `ha-area-picker`, `ha-device-picker`, `ha-entity-picker` and `ha-combo-box`:

```diff
@ -0,0 +1,194 @@
diff --git a/src/components/device/ha-device-picker.ts b/src/components/device/ha-device-picker.ts
index b387be111..d93e717b2 100644
--- a/src/components/device/ha-device-picker.ts
+++ b/src/components/device/ha-device-picker.ts
@@ -221,12 +221,14 @@ export class HaDevicePicker extends SubscribeMixin(LitElement) {
     }
   );
 
-  public open() {
-    this.comboBox?.open();
+  public async open() {
+    await this.updateComplete;
+    await this.comboBox?.open();
   }
 
-  public focus() {
-    this.comboBox?.focus();
+  public async focus() {
+    await this.updateComplete;
+    await this.comboBox?.focus();
   }
 
   public hassSubscribe(): UnsubscribeFunc[] {
@@ -246,7 +248,7 @@ export class HaDevicePicker extends SubscribeMixin(LitElement) {
   protected updated(changedProps: PropertyValues) {
     if (
       (!this._init && this.devices && this.areas && this.entities) ||
-      (changedProps.has("_opened") && this._opened)
+      (this._init && changedProps.has("_opened") && this._opened)
     ) {
       this._init = true;
       (this.comboBox as any).items = this._getDevices(
@@ -262,9 +264,6 @@ export class HaDevicePicker extends SubscribeMixin(LitElement) {
   }
 
   protected render(): TemplateResult {
-    if (!this.devices || !this.areas || !this.entities) {
-      return html``;
-    }
     return html`
       <ha-combo-box
         .hass=${this.hass}
diff --git a/src/components/entity/ha-entity-picker.ts b/src/components/entity/ha-entity-picker.ts
index ef3c6ee59..056918602 100644
--- a/src/components/entity/ha-entity-picker.ts
+++ b/src/components/entity/ha-entity-picker.ts
@@ -107,16 +107,14 @@ export class HaEntityPicker extends LitElement {
 
   @query("ha-combo-box", true) public comboBox!: HaComboBox;
 
-  public open() {
-    this.updateComplete.then(() => {
-      this.comboBox?.open();
-    });
+  public async open() {
+    await this.updateComplete;
+    await this.comboBox?.open();
   }
 
-  public focus() {
-    this.updateComplete.then(() => {
-      this.comboBox?.focus();
-    });
+  public async focus() {
+    await this.updateComplete;
+    await this.comboBox?.focus();
   }
 
   private _initedStates = false;
diff --git a/src/components/ha-area-picker.ts b/src/components/ha-area-picker.ts
index 6a56868bd..270a7ed8c 100644
--- a/src/components/ha-area-picker.ts
+++ b/src/components/ha-area-picker.ts
@@ -116,16 +116,14 @@ export class HaAreaPicker extends SubscribeMixin(LitElement) {
     ];
   }
 
-  public open() {
-    this.updateComplete.then(() => {
-      this.comboBox?.open();
-    });
+  public async open() {
+    await this.updateComplete;
+    await this.comboBox?.open();
   }
 
-  public focus() {
-    this.updateComplete.then(() => {
-      this.comboBox?.focus();
-    });
+  public async focus() {
+    await this.updateComplete;
+    await this.comboBox?.focus();
   }
 
   private _getAreas = memoizeOne(
@@ -290,7 +288,7 @@ export class HaAreaPicker extends SubscribeMixin(LitElement) {
   protected updated(changedProps: PropertyValues) {
     if (
       (!this._init && this._devices && this._areas && this._entities) ||
-      (changedProps.has("_opened") && this._opened)
+      (this._init && changedProps.has("_opened") && this._opened)
     ) {
       this._init = true;
       (this.comboBox as any).items = this._getAreas(
@@ -308,9 +306,6 @@ export class HaAreaPicker extends SubscribeMixin(LitElement) {
   }
 
   protected render(): TemplateResult {
-    if (!this._devices || !this._areas || !this._entities) {
-      return html``;
-    }
     return html`
       <ha-combo-box
         .hass=${this.hass}
diff --git a/src/components/ha-combo-box.ts b/src/components/ha-combo-box.ts
index fa328e4dd..5289f1465 100644
--- a/src/components/ha-combo-box.ts
+++ b/src/components/ha-combo-box.ts
@@ -1,5 +1,6 @@
 import "@material/mwc-list/mwc-list-item";
 import { mdiClose, mdiMenuDown, mdiMenuUp } from "@mdi/js";
+import { ComboBoxLitRenderer, comboBoxRenderer } from "@vaadin/combo-box/lit";
 import "@vaadin/combo-box/theme/material/vaadin-combo-box-light";
 import type {
   ComboBoxLight,
@@ -16,13 +17,13 @@ import {
   PropertyValues,
   TemplateResult,
 } from "lit";
-import { ComboBoxLitRenderer, comboBoxRenderer } from "@vaadin/combo-box/lit";
 import { customElement, property, query } from "lit/decorators";
 import { ifDefined } from "lit/directives/if-defined";
 import { fireEvent } from "../common/dom/fire_event";
 import { HomeAssistant } from "../types";
 import "./ha-icon-button";
 import "./ha-textfield";
+import type { HaTextField } from "./ha-textfield";
 
 registerStyles(
   "vaadin-combo-box-item",
@@ -108,18 +109,19 @@ export class HaComboBox extends LitElement {
 
   @query("vaadin-combo-box-light", true) private _comboBox!: ComboBoxLight;
 
+  @query("ha-textfield", true) private _inputElement!: HaTextField;
+
   private _overlayMutationObserver?: MutationObserver;
 
-  public open() {
-    this.updateComplete.then(() => {
-      this._comboBox?.open();
-    });
+  public async open() {
+    await this.updateComplete;
+    this._comboBox?.open();
   }
 
-  public focus() {
-    this.updateComplete.then(() => {
-      this._comboBox?.inputElement?.focus();
-    });
+  public async focus() {
+    await this.updateComplete;
+    await this._inputElement?.updateComplete;
+    this._inputElement?.focus();
   }
 
   public disconnectedCallback() {
diff --git a/src/components/ha-target-picker.ts b/src/components/ha-target-picker.ts
index 0f6ee2f41..ec466335c 100644
--- a/src/components/ha-target-picker.ts
+++ b/src/components/ha-target-picker.ts
@@ -21,6 +21,7 @@ import { fireEvent } from "../common/dom/fire_event";
 import { ensureArray } from "../common/ensure-array";
 import { computeDomain } from "../common/entity/compute_domain";
 import { computeStateName } from "../common/entity/compute_state_name";
+import { nextRender } from "../common/util/render-status";
 import {
   AreaRegistryEntry,
   subscribeAreaRegistry,
@@ -251,10 +252,8 @@ export class HaTargetPicker extends SubscribeMixin(LitElement) {
   private async _showPicker(ev) {
     this._addMode = ev.currentTarget.type;
     await this.updateComplete;
-    setTimeout(() => {
-      this._inputElement?.open();
-      this._inputElement?.focus();
-    }, 0);
+    await this._inputElement?.focus();
+    this._inputElement?.open();
   }
 
   private _renderChip(
   ```

Reply:
I can push this if you want

Reply:
Exactly what I was thinking 👍 

Although it may not work because, as I pointed out, `mwc-textfield` does not await its own update on focus.

Reply:
I fixed that by awaiting it myself in `ha-combo-box` :-)

Reply:
Thanks! I can look into it in about 8 hours.

Reply:
@bramkragten , I've implemented your code. It's a much better change for sure!